pkg/controllers/user/cis/clusterScanHandler.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	projv3 "github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type cisScanHandler struct {
+	clusterLister                v3.ClusterLister
+	projectLister                v3.ProjectLister
+	mgmtCtxClusterClient         v3.ClusterInterface
+	mgmtCtxProjClient            v3.ProjectInterface
+	mgmtCtxAppClient             projv3.AppInterface
+	mgmtCtxTemplateVersionClient v3.CatalogTemplateVersionInterface
+	mgmtCtxClusterScanClient     v3.ClusterScanInterface
+	userCtx                      *config.UserContext
+	userCtxNSClient              rcorev1.NamespaceInterface
+	clusterNamespace             string
+	systemAccountManager         *systemaccount.Manager
+	configMapsClient             rcorev1.ConfigMapInterface
+}
+
+func (csh *cisScanHandler) Create(cs *v3.ClusterScan) (runtime.Object, error) {
+	logrus.Debugf("cisScanHandler: Create: %+v", cs)
+	var err error
+	cluster, err := csh.clusterLister.Get("", cs.Spec.ClusterID)
+	if err != nil {
+		return cs, fmt.Errorf("cisScanHandler: Create: error listing cluster %v: %v", cs.ClusterName, err)
+	}
+	if !v3.ClusterConditionReady.IsTrue(cluster) {
+		return cs, fmt.Errorf("cisScanHandler: Create: cluster %v not ready", cs.ClusterName)
+	}
+	if !v3.ClusterScanConditionCreated.IsTrue(cs) {
+		logrus.Infof("cisScanHandler: Create: deploying helm chart")
+		// Deploy the system helm chart
+		if err := csh.deployApp(cs.Spec.ClusterID, cs.Name); err != nil {
+			return cs, fmt.Errorf("cisScanHandler: Create: error deploying app: %v", err)
+		}
+		v3.ClusterScanConditionCreated.True(cs)
+		v3.ClusterScanConditionCompleted.Unknown(cs)
+
+		cs, err = csh.mgmtCtxClusterScanClient.Update(cs)
+		if err != nil {
+			return cs, fmt.Errorf("cisScanHandler: Create: error updating cs: %v error: %v", cs.Name, err)
+		}
+	}
+	return cs, nil
+}
+
+func (csh *cisScanHandler) Remove(cs *v3.ClusterScan) (runtime.Object, error) {
+	logrus.Debugf("cisScanHandler: Remove: %+v", cs)
+	// Delete the configmap associated with this scan
+	err := csh.configMapsClient.Delete(cs.Name, &metav1.DeleteOptions{})
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			return cs, fmt.Errorf("cisScanHandler: Remove: error deleting cm=%v", cs.Name)
+		}
+		return cs, nil
+	}
+
+	if err := csh.deleteApp(cs.Name); err != nil {
+		return nil, fmt.Errorf("cisScanHandler: Remove: error deleting app: %v", err)
+	}
+
+	cluster, err := csh.clusterLister.Get("", csh.clusterNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("cisScanHandler: Remove: error getting cluster %v", err)
+	}
+
+	if owner, ok := cluster.Annotations[RunCISScanAnnotation]; ok && owner == cs.Name {
+		updatedCluster := cluster.DeepCopy()
+		delete(updatedCluster.Annotations, RunCISScanAnnotation)
```

Comment:
there might be a chance when the update fails. And then it will never be retried because we return on the line 70.